### ISSUE TITLE
Tensorflow Lite exhibits longer inference time when build with Android NN API on Google Pixel 1

### ISSUE THREAD

**Issue by pointcontrols** [tensorflow/tensorflow#15554]
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
'v1.3.0-rc1-6055-gfdf34a8', '1.4.0'
- **Python version**: 
2.7.12
- **Bazel version (if compiling from source)**:
0.8.0
- **GCC/Compiler version (if compiling from source)**:
GCC 5.4.0-6ubuntu1~16.04.5
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
When enabling the NN API usage (edit interpreter.cc) one would expect that due to the HW acceleration, inferece times would be shorter. However, exactly the opposite happens. I tried this using the included demo app, using mobilenet_quant_v1_224.tflite and a custom (not quantized) model that I converted to tflite. In the mobilenet case, inference takes around 80ms without NNAPI, and ~100ms with NNAPI . My custom model, which sadly I cannot share , takes 40ms without , and ~90ms with NNAPI.

Please note that I have also tested this with FAST_SINGLE_ANSWER  and SUSTAINED_SPEED  NNAPI preference settings. There was no significant change in inference times.

My Pixel Build number is: OPM1.171019.011

### Source code / logs
--- a/tensorflow/contrib/lite/interpreter.cc
+++ b/tensorflow/contrib/lite/interpreter.cc
@@ -51,7 +51,7 @@ Interpreter::Interpreter(ErrorReporter* error_reporter)
   tensors_.reserve(kSlotsToReserve);
   nodes_and_registration_.reserve(kSlotsToReserve);
   next_allocate_node_id_ = 0;
-  UseNNAPI(false);
+  UseNNAPI(true);

**Comment by miaowang14** [comment]
Thanks for reporting this.

Pixel 2016 currently does not have a hardware driver for NN API. So your model is actually running on CPU, plus some overhead for model initialization.

Another source of overhead is for input and output tensors. If the buffer is not already created as shared memory, an copy might occur during invocation. We will keep improving the integration to make it more efficient.

**Comment by pointcontrols** [comment]
Thanks. Which devices do have a HW acceleration support?

**Comment by pointcontrols** [comment]
@miaowang14 Can you please share, is there a device available today that does have a working NN API hw acceleration ? I have a pretty big decision to make based on this information. Thanks!

**Comment by mkassner** [comment]
@miaowang14 @aselle is there a list of hardware/phones where TF actually uses GPU or DSP? I think TF light is a great idea but without hardware support I m not sure how much adoption it will see.

If you know of at least one supported phone we can get started developing with that! I would very much like to work with TF as opposed to a custom solution but we will need HW support :-)

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mkassner** [comment]
@miaowang14  just a gentle reminder. Any update you could share? Or at least an indication  if this issue is of priority?

**Comment by jeanlucbr** [comment]
mkassner, there are many devices coming out with hardware acceleration.  I'll see if we can get permission to share a list of devices.

Thanks for looking into TFLite and the NNAPI.

**Comment by mkassner** [comment]
@jeanlucbr  a list of supported devices would be great! Will some of the devices be released this year?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jeanlucbr** [comment]
@mkassner Unfortunately, I could not get permission from the vendors to reveal the devices, as many are unannounced.  Keep your eyes open, I expect multiple announcements soon.  It's worth the wait; the acceleration is significant.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Arjuna197** [comment]
So does the Pixel 2 XL also does not feature a hardware driver for nnapi?

**Comment by rohan100jain** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by aiff22** [comment]
We are recently doing phone ai benchmarking and checking the support of NNAPI on various android devices. The situation is now quite sad: in brief, only Huawei phones with Kirin's 970 NPU have drivers for it, while other devices are still lacking them. In general, the situation with TFLite/NNAPI is even more curious:

1) TF Lite is ~1.5-2 times slower and consumes ~1.5-2 times more RAM than TF Mobile.
2) If drivers for NNAPI are not available, this slows TFLite down by another 1.5-2 times.
3) NNAPI supports only a subset of TF Lite operations, therefore the majority of models will crush after running with nnapi acceleration - even Inception-v1 and MobileNets are currently not supported, not to mention more complicated models.

We are trying to keep track of the situation and are publishing the current performance of tflite/tfmobile on various devices on our webpage: http://ai-benchmark.com/ranking. The phones supporting nnapi are denoted by green (currently it's only huawei), and as soon as new devices with acceleration support are released they will be added to this rating (btw, you can also find the running time and memory requirements for several classification and image-to-image translation networks on this page). If you need to check whether your phone supports nnapi, you can also run the app on it and at the end you will either get a message that it is supported, or a warning that it lacks the drivers.

Hope this helps.

**Comment by shashishekhar** [comment]
@aiff22 
Thank you for the report.

Which model are you benchmarking on these devices.

Are you using the Java API?
 - TF Lite being 1.5- 2 times slower seems unlikely, I want to make sure if you are setting the number of threads correctly. We recently added this to our Java API, please give it a try.
- Java API is slower if you are not using ByteBuffer, there is a warning in the Javadoc about this:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java#L134
Make sure you are using the right format for inputs.

If you are using C++, can you use our newly added benchmarking binary and see if the numbers are same:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/lite/tools/benchmark
and report detailed numbers on which ops seem slower.

Tensorflow mobile has a similar binary in:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/benchmark

**Comment by aiff22** [comment]
@shashishekhar 

Thank you for a fast reply. Yes, we are using Java API (with ByteBuffer), setting #threads wasn't available just recently, so it was set to some default number. We will try to play around it now and will let you know whether this will lead to better performance. However, most likely this won't solve the issue with RAM, and we guess that those two might be connected to each other.

We benchmarked several models, including Inception-v1/v3, Inception-ResNet-v1, SRCNN, ResNet, VGG-19, ICNet and some other architectures (including both image classification and image-to-image translation nets). The difference between inference times for TFMobile/TFLite/TFLite+NNAPI was quite stable among all networks. In particular, this also applies for SRCNN that consists of two ops: conv and relu, and thus can be considered as a simplest baseline.

**Comment by aiff22** [comment]
We tried changing the number of threads, but this didn't solve the issue. Moreover, it seems that setNumThreads method has absolutely no affect on TF Lite's performance - for values of 1, 2, 4, 8, 100 and 1000 the results were identical (up to several ms).

Here is more information regarding the performance of TF Mobile, TF Lite and TF Lite + NNAPI. We've just tested SRCNN (image deblurring, 200x200px) and VGG-19 (image super-res, 192x192px) networks on two devices - Nexus 5X with Android 8.1 and Razer Phone with Android 7.1, below are the results:

Nexus 5X, TF Mobile: SRCNN - 0.8s, VGG-19 - 4.5s
Nexus 5X, TF Lite: SRCNN - 1.8s, VGG-19 - 8s
Nexus 5X, TF Lite + NNAPI: SRCNN - 2.7s, VGG-19 - 15.7s

Razer Phone, TF Mobile: SRCNN - 0.36s, VGG-19 - 1.8s
Razer Phone, TF Lite: SRCNN - 0.71s, VGG-19 - 3.3s

As mentioned before, changing the number of threads didn't lead to any improvements.
